### PR TITLE
Reducing duplicates in Phase2 tracking reconstruction (with some minor effects on phase0/1 tracking)

### PR DESCRIPTION
**Pull request by ebrondol**
With this PR, I am importing the PR#3051 done in SLHC by @cerati .
A first comparison can be seen [here](http://ebrondol.web.cern.ch/ebrondol/Migration81X/MTV_TTbar_81xvsSLHCDEV_fix/plots_ootb/effandfake1.pdf). 
Since this part is touching the run2 reconstruction, I have already run the most important WFs and check that everything produce the same results. Still, a more detail test analysis is for sure needed.

@VinInn @rovere

### CODE REVIEW

File: RecoTracker/CkfPattern/plugins/GroupedCkfTrajectoryBuilder.cc [hunk 1]
```diff
@@ -848,11 +848,11 @@ GroupedCkfTrajectoryBuilder::rebuildSeedingRegion(const TrajectorySeed&seed,
...
-    if ( it->measurements().size()<=startingTraj.measurements().size() ) {
-      rebuiltTrajectories.push_back(std::move(*it));
-      LogDebug("CkfPattern")<< "RebuildSeedingRegion skipped as in-out trajectory does not exceed seed size.";
-      continue;
-    }
+    //if ( it->measurements().size()<=startingTraj.measurements().size() ) {
+    //  rebuiltTrajectories.push_back(std::move(*it));
+    //  LogDebug("CkfPattern")<< "RebuildSeedingRegion skipped as in-out trajectory does not exceed seed size.";
+    //  continue;
+    //}
```
Comment (slava77) [inline RecoTracker/CkfPattern/plugins/GroupedCkfTrajectoryBuilder.cc]: shouldn't this and the other one be made configurable?
That's unless we never ever make it into this branch in current tracking.

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoTracker/CkfPattern/plugins/GroupedCkfTrajectoryBuilder.cc (hunk 1 above)
Comment (ebrondol) [inline RecoTracker/CkfPattern/plugins/GroupedCkfTrajectoryBuilder.cc]: As far as I searched, this change was never made in the running tracking.

**Comment by slava77** [comment]
@cmsbuild please test
hopefully not DAS problems this time

**Comment by makortel** [comment]
FWIW, I had already the MTV plots for phase1 (1000 ttbar+PU events)
https://mkortela.web.cern.ch/mkortela/tracking/validation/CMSSW_8_1_0_pre5_v17/
(blue is the phase1 tracking as in 810pre7, and red is with the changes in this PR in addition; although run in 810pre5 with 810pre4 RAW). In short, there are tiny changes that look mostly be caused changes in track pT.

**Comment by ebrondol** [comment]
Thanks @makortel 
Hi @slava77 I had a look at the "Comparison with the baseline" and for all the samples I have not noticed any difference. Please, can you be more specific on where you have seen the differences and for which samples you would suggest to start to have the MTV?

**Comment by slava77** [comment]
On 6/29/16 9:57 AM, ebrondol wrote:
> Thanks @makortel <https://github.com/makortel>
> Hi @slava77 <https://github.com/slava77> I had a look at the "Comparison
> with the baseline" and for all the samples I have not noticed any
> difference. Please, can you be more specific on where you have seen the
> differences and for which samples you would suggest to start to have the
> MTV?
> 

jenkins shows differences in 1000.0 and 136.731
It looks like mainly the algo changes (but these are low stat tests)
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_8_1_X_2016-06-27-1100+14994/14643/validateJR/all_OldVSNew_RunMinBias2011Awf1000p0/all_OldVSNew_RunMinBias2011Awf1000p0c_recoTracks_generalTracks__RECO_obj_algo.png



> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/14994#issuecomment-229283932>, or
> mute the thread
> <https://github.com/notifications/unsubscribe/AEdcbkoQFTfoRiasQCK966XoxRFi4A9uks5qQiVpgaJpZM4I_Kfi>.
> 

-

**Comment by VinInn** [comment]
baseline is pre7 or similar
http://innocent.home.cern.ch/innocent/RelVal/pu35_pr14994/plots_highPurity/effandfake1.pdf
http://innocent.home.cern.ch/innocent/RelVal/pu35_pr14994/pu_plots_summary.html

**Comment by ebrondol** [comment]
Hi @slava77, the results produced by @VinInn are enough?

**Comment by slava77** [comment]
@ebrondol 
I didn't get to this yet. Likely, tomorrow.

File: RecoTracker/CkfPattern/plugins/GroupedCkfTrajectoryBuilder.cc (hunk 1 above)
Comment (slava77) [inline RecoTracker/CkfPattern/plugins/GroupedCkfTrajectoryBuilder.cc]: If this code is not needed, please remove.
If for some reason it's needed here commented out, please add a clear comment why.
The same for a change in line 1021

**Comment by slava77** [comment]
@ebrondol 
please improve the title of the PR.
It should reflect that run2/current tracking behavior changes as well (may add "minor effects on run1/2 tracking")

File: RecoTracker/CkfPattern/plugins/GroupedCkfTrajectoryBuilder.cc (hunk 1 above)
Comment (ebrondol) [inline RecoTracker/CkfPattern/plugins/GroupedCkfTrajectoryBuilder.cc]: I do not have a strong opinion on that, @VinInn or @rovere do we want to keep it for a future reference or we can get rid of it? @cerati did you keep it in your PR at that time for a specific reason?

**Comment by slava77** [comment]
As expected, there are significant changes mainly in 2023 workflows.

most typical change in run2 workflow is a slight change in the algorithm or quality mask (in the workflows with a change a pf charged hadron would be lost)
e.g.  200 events in doubleMu 2016B
all_sign735vsorig_rundoublemuon2016bwf136p723c_recotracks_generaltracks__rereco_obj_qualitymask

2023 ttbar 10624:
all_sign735vsorig_ttbar13tev2023wf10624p0c_recopfcandidates_particleflow__reco_obj_eta12
less tracks from higher iterations
all_sign735vsorig_ttbar13tev2023wf10624p0c_recotracks_generaltracks__reco_obj_algo
all_sign735vsorig_ttbar13tev2023wf10624p0c_recotracks_generaltracks__reco_obj_found

similarly in 2023 QCD 600-800 (wf 10626).

Beyond this, I'm relying on the MTV plots posted earlier at PR submission
https://github.com/cms-sw/cmssw/pull/14994#issue-162468461


Other than the issue in 
https://github.com/cms-sw/cmssw/pull/14994#discussion_r69607802
this looks good to go

File: RecoTracker/CkfPattern/plugins/GroupedCkfTrajectoryBuilder.cc (hunk 1 above)
Comment (VinInn) [inline RecoTracker/CkfPattern/plugins/GroupedCkfTrajectoryBuilder.cc]: delete

**Comment by ebrondol** [comment]
Removed.

**Comment by VinInn** [comment]
@cmsbuild , please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/14994	5fcf103
- code change is OK
- jenkins tests (now include 2023 workflows!) confirm expected behavior
- results from tests with more events were posted earlier https://github.com/cms-sw/cmssw/pull/14994#issuecomment-230832286 (the last commit removed only the initially commented out lines)

**Comment by ebrondol** [comment]
@kpedro88 just for your info

**Pull request merged**